### ISSUE TITLE
Speedup in RepeatedAffineComponent

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#441]
@galv, if you have time you can do this too [in the 'chain' branch].

It looks like the function AddMatMatBatched, and the underlying sgemmBatched, is not as optimized as it could be.  This makes the RepeatedAffineComponent slower than it should be, which is affecting the speed of some of my experiments.

There is a way we can optimize this not too painfully, by turning everything into a single matrix multiply.   This will make the code a lot simpler too.  Basically it will require two temporary matrices, one for the input and one for the output [or one for the input-derivative and one for the output-derivative, in backprop].  These matrices will have a lot more rows than the original matrices-- the number of rows will be greater by a factor of the group-size.  To avoid wasting too much memory due to the padding of the stride, you might want to experiment with making them both sub-matrices of a single larger matrix with the same number of rows, but a larger number of columns- or at least make it easy in the code to try both, since experimentation may be easier for me.  [Note: if the absolute amount of memory becomes too large we can reduce with the 'max-rows-process' option of CompositeComponent].

There is a function called Tensor3dCopy, which is defined in ctc/cctc-tombstone.cc in the following commit on my personal repo (e.g. accessible if you are currently on the jesus branch), which you will find useful in doing the data copying.  You can access this as follows:
git checkout c83118672dc5e1ddbdece094fcb5b1fe5422d78d~1 -b tensor3dcopy

This can be moved to cu-matrix.{h,cc}.
There is some CUDA code you'll need for this, which is located in cctc-kernels.cu [can be moved to cu-kernels.h], and a declaration located in cctc-kernels-ansi.h [move to cu-kernels-ansi.h] and in cctc-kernels.h [move to cu-kernels.h].

**Comment by galv** [comment]
I've seen a trick like this used in implementing LSTMs actually, for calculating each gate all at once, in a single gemm call. I also read the CUBLAS docs on the batched gemm call; I agree it's wrong for this component. I had no idea how large the dimensions would be. Sorry. Anyway:

First of all, so we're on the same page:
```
group-size == input_dim / _num_repeats
```
right? Here, **input_dim** is the input feature dimension to the component itself. **output_dim**, which I introduce later similar denote the output feature dimension of this component (the hidden feature dimension). **N** is the number of elements in a batch. In the following, "*" denotes multiplication, while "X" denotes what we typically mean when we say a matrix M X N (M-by-N).

~~For forward propagation, we want a 3-tensor of dimensionality **group_size** X **N** X **input_dim**, but we only have matrices, so we emulate this as a **group_size** * **N** X **input_dim** matrix. This is what you mean by "the number of rows will be greater by a factor of the group-size." Correct?~~

Never mind, what I believe we need is a 3-tensor of dimensionality **N** X **_num_repeats** X **group-size**. We will emulate this as a matrix as a **N** * _num_repeats * group-size matrix. The rightmost dimension would then be multiplied by the single _linear_params matrix (of dimension **group-size** X **output-group-size**. output-group-size is defined as **output_dim** / **_num_repeats**). This is the "temporary" matrix, of dimension reshaping from **N** * **_num_repeats** X **output-group-size**. The output is then a reshaping of the temporary matrix to **N** X **_num_repeats** * **output-group-size**. Note that this second dimension is defined to be **output_dim**. I believe I see how tensor3dcopy() can do what we need.

However, I'm not particularly sure why we need a second temporary matrix, based on my discussion above. **I know this is a long response, so if you respond to anything, respond to the previous sentence.**

I am not sure what you mean here: "To avoid wasting too much memory due to the padding of the stride". I don't know the details of how CUDA does padding, but in OpenCV the only purpose of padding is to make sure that memory accesses are properly aligned. I don't see how padding could "waste" more than a few bytes of memory. These are row-major matrices, and the number of columns (i.e., row length) is quite large.

Finally, it is possible that we don't actually have to copy using Tensor3dCopy. I have to think about this some more. Unfortunately, this totally break the matrix abstraction, but I will investigate. I think this is possible from my playing with torch7's tensor type, which chooses to change strides often to avoid memory copies at all costs. Obviously, some messing with the stride may reduce data locality. I need to think it through.

Tell me if something doesn't make sense.

**Comment by galv** [comment]
Also, have you done #438 on a personal branch, or have you simply come to the conclusive that cublasSgemm is slow by viewing training logs?

**Comment by danpovey** [comment]
i have not tested (#438).
2nd matrix is needed because there is an input and an output in the core
sgemm call.
padding is an issue because the cublas library pads a lot, I think 256
bytes.
Dan

**Comment by galv** [comment]
Regarding 2nd matrix: yes, of course. duh.

On Fri, Jan 8, 2016 at 2:40 AM, Daniel Povey <notifications@github.com>
wrote:

> i have not tested (#438).
> 2nd matrix is needed because there is an input and an output in the core
> sgemm call.
> padding is an issue because the cublas library pads a lot, I think 256
> bytes.
> Dan
>
> On Thu, Jan 7, 2016 at 11:33 PM, Daniel Galvez <notifications@github.com>
> wrote:
>
> > I've seen a trick like this used in implementing LSTMs actually, for
> > calculating each gate all at once, in a single gemm call. I also read the
> > CUBLAS docs on the batched gemm call; I agree it's wrong for this
> > component. I had no idea how large the dimensions would be. Sorry.
> Anyway:
> >
> > First of all, so we're on the same page:
> >
> > group-size == input_dim / _num_repeats
> >
> > right? Here, *input_dim* is the input feature dimension to the component
> > itself. *output_dim*, which I introduce later similar denote the output
> > feature dimension of this component (the hidden feature dimension). *N*
> > is the number of elements in a batch. In the following, "*" denotes
> > multiplication, while "X" denotes what we typically mean when we say a
> > matrix M X N (M-by-N).
> >
> > For forward propagation, we want a 3-tensor of dimensionality
> *group_size*
> > X *N* X *input_dim*, but we only have matrices, so we emulate this as a
> > *group_size* * *N* X *input_dim* matrix. This is what you mean by "the
> > number of rows will be greater by a factor of the group-size." Correct?
> >
> > Never mind, what I believe we need is a 3-tensor of dimensionality *N* X
> > *_num_repeats* X *group-size*. We will emulate this as a matrix as a *N*
> > * _num_repeats * group-size matrix. The rightmost dimension would then be
> > multiplied by the single _linear_params matrix (of dimension *group-size*
> > X *output-group-size*. output-group-size is defined as *output_dim* /
> > *_num_repeats*). This is the "temporary" matrix, of dimension reshaping
> > from *N* * *_num_repeats* X *output-group-size*. The output is then a
> > reshaping of the temporary matrix to *N* X *_num_repeats* *
> > *output-group-size*. Note that this second dimension is defined to be
> > *output_dim*. I believe I see how tensor3dcopy() can do what we need.
> >
> > However, I'm not particularly sure why we need a second temporary matrix,
> > based on my discussion above. *I know this is a long response, so if you
> > respond to anything, respond to the previous sentence.*
> >
> > I am not sure what you mean here: "To avoid wasting too much memory due
> to
> > the padding of the stride". I don't know the details of how CUDA does
> > padding, but in OpenCV the only purpose of padding is to make sure that
> > memory accesses are properly aligned. I don't see how padding could
> "waste"
> > more than a few bytes of memory. These are row-major matrices, and the
> > number of columns (i.e., row length) is quite large.
> >
> > Finally, it is possible that we don't actually have to copy using
> > Tensor3dCopy. I have to think about this some more. Unfortunately, this
> > totally break the matrix abstraction, but I will investigate. I think
> this
> > is possible from my playing with torch7's tensor type, which chooses to
> > change strides often to avoid memory copies at all costs. Obviously, some
> > messing with the stride may reduce data locality. I need to think it
> > through.
> >
> > Tell me if something doesn't make sense.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/issues/441#issuecomment-169917989>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/issues/441#issuecomment-169919075>.
>



-- 
Daniel Galvez
danieltgalvez.com

**Comment by galv** [comment]
I chose to implement this by adding a new optional boolean "fast-mode" to the config line; this is so we can more easily see if this really is faster simply by changing a configuration.

Also, there's no way to simply change our view of the striding to avoid using Tensor3dCopy() because of the padding at the end of each row. I looked into Torch7. The reason why they can do this is because they never use cudaMallocPitch; instead they always use cudaMalloc (even for 2d matrices and larger tensors), and therefore never have padding.

Anyway, I need some advice regrading a design decision.

The two extra matrices: I have made them instance variables of the component's class. The idea is to avoid freeing and mallocing after every backprop call and every propagate call. (Their sizes make it so that they can be used as "scratch" space in both functions.) They are modified by the Propagate() and Backprop() functions, but I declare them mutable so that the functions are logical const (basically, bitwise const from user's point-of-view, though not bitwise const from

1) Is it true that every Backprop call is preceded by a Propagate call? Just need to make sure.
2) Is it okay to keep these "scratch" space matrices allocated? Just want to know if we're squeezed for space.
3) Do you also want NaturalGradientRepeatedAffineComponent implemented this same way? (i.e., one gemm call instead of multiple gemm calls.) I am trying to make backprop do the minimal amount of computation, and want to keep this in mind.
4) Does batch size ever change?

Propagate() is now implemented. I've tested it with the old Backprop implementation. It works.

**Comment by danpovey** [comment]
>
> I chose to implement this by adding a new optional boolean "fast-mode" to
> the config line; this is so we can more easily see if this really is faster
> simply by changing a configuration.
>
> Also, there's no way to simply change our view of the striding to avoid
> using Tensor3dCopy() because of the padding at the end of each row. I
> looked into Torch7. The reason why they can do this is because they never
> use cudaMallocPitch; instead they always use cudaMalloc (even for 2d
> matrices and larger tensors), and therefore never have padding.
>
OK.  Apparently the stride used to be very important for older hardware but
is becoming less and less critical- we could investigate changing this at
some point but it would be a big change.

Anyway, I need some advice regrading a design decision.
>
> The two extra matrices: I have made them instance variables of the
> component's class. The idea is to avoid freeing and mallocing after every
> backprop call and every propagate call. (Their sizes make it so that they
> can be used as "scratch" space in both functions.) They are modified by the
> Propagate() and Backprop() functions, but I declare them mutable so that
> the functions are logical const (basically, bitwise const from user's
> point-of-view, though not bitwise const from
>
Don't do this.  It goes against the use of 'const' in the class and would
break multi-threaded updates.  Anyway freeing and mallocing is not slow at
all as long as you initialize them with kUndefined; we have caching code
which prevents it from doing anything heavyweight unless we're almost out
of memory.


> 1) Is it true that every Backprop call is preceded by a Propagate call?
> Just need to make sure.
>
You can't assume this, and in fact the same component may be used for
multiple matrices in the computation.

> 2) Is it okay to keep these "scratch" space matrices allocated? Just want
> to know if we're squeezed for space.
>
no...

Dan

**Comment by galv** [comment]
On Sat, Jan 9, 2016 at 6:07 PM, Daniel Povey <notifications@github.com>
wrote:

> >
> > I chose to implement this by adding a new optional boolean "fast-mode" to
> > the config line; this is so we can more easily see if this really is
> faster
> > simply by changing a configuration.
> >
> > Also, there's no way to simply change our view of the striding to avoid
> > using Tensor3dCopy() because of the padding at the end of each row. I
> > looked into Torch7. The reason why they can do this is because they never
> > use cudaMallocPitch; instead they always use cudaMalloc (even for 2d
> > matrices and larger tensors), and therefore never have padding.
> >
> OK. Apparently the stride used to be very important for older hardware but
> is becoming less and less critical- we could investigate changing this at
> some point but it would be a big change.
>
> May not be worth it anyhow; reviewing the matrix will reduce data locality
for the cache.

> Anyway, I need some advice regrading a design decision.
> >
> > The two extra matrices: I have made them instance variables of the
> > component's class. The idea is to avoid freeing and mallocing after every
> > backprop call and every propagate call. (Their sizes make it so that they
> > can be used as "scratch" space in both functions.) They are modified by
> the
> > Propagate() and Backprop() functions, but I declare them mutable so that
> > the functions are logical const (basically, bitwise const from user's
> > point-of-view, though not bitwise const from
> >
> Don't do this. It goes against the use of 'const' in the class and would
> break multi-threaded updates. Anyway freeing and mallocing is not slow at
> all as long as you initialize them with kUndefined; we have caching code
> which prevents it from doing anything heavyweight unless we're almost out
> of memory.
>
>
> > 1) Is it true that every Backprop call is preceded by a Propagate call?
> > Just need to make sure.
> >
> You can't assume this, and in fact the same component may be used for
> multiple matrices in the computation.
>
> > 2) Is it okay to keep these "scratch" space matrices allocated? Just want
> > to know if we're squeezed for space.
> >
> no...
>
> Dan
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/issues/441#issuecomment-170291753>.
>

Thanks!

-- 
Daniel Galvez
danieltgalvez.com
